microsoft/fast
# chore(fast-todo-app): setup custom elements language server

# Pull Request

## 📖 Description


Over the last six months we've been working on an LSP plugin for the `tsserver` which adds custom elements support in tagged template literals. The LSP is build to be extensible for different custom elements dialects, but we've added support for Microsoft FAST ourselves. 
This PR installs the LSP in the `todo-app` example repo as a showcase. This isn't necessarily the best showcase of the LSP due to the fact that this app doesn't use many custom elements, or any from dependencies. I am keen to install it more globally in the repo, or across more apps, dependening on whether you want this adding at all. 

![demo](https://github.com/microsoft/fast/assets/43502076/23643e89-d2da-4f17-97a5-eac71b83b870)


### 🎫 Issues



## 👩‍💻 Reviewer Notes



## 📑 Test Plan


The way to test this depends on your IDE setup, and the way it works with the LSP. You can see in depth info for this [here](https://www.npmjs.com/package/@genesiscommunitysuccess/custom-elements-lsp#vscode), but I assume you're using vscode so will provide some quick instrucitons below.
```
1. Check out the branch and use yarn to update the installed dependencies.
2. Navigate to the `todo-app` project directory
3. `$ mkdir .vscode && echo '{ "typescript.tsdk": "../../node_modules/typescript/lib" }' >> .vscode/settings.json` ⚠️ if you've already got project specific settings in this file for your editor then you can manually set the config in your existing file instead.
4. Open vscode at this todo app project level `$ code .`
5. Open a javascript/typescript file and use the command palette to `Select typescript version` and choose the local version.
6. The LSP should now be enabled. You will only see diagnostics information after first interacting with the file
```

You can observe in `todo-app.template.ts` if you type a `<` you'll get a suggestion for `<todo-form>`

## ✅ Checklist

### General



- [ ] I have included a change request file using `$ yarn change` - [Contributing guidelines](https://github.com/microsoft/fast/blob/daff330978c2bbc961e729ee7f285ab426bf0beb/CONTRIBUTING.md#change-files) state this is only necessary if you're changing files in the `packages/*` directory, which this PR is not.
- [ ] I have added tests for my changes. - N/A
- [ ] I have updated the project documentation to reflect my changes. - N/A
- [X] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific





None of these

## ⏭ Next Steps

Happy to discuss or answer any questions you may have!
